crates/bevy_ecs/src/ptr.rs

```diff
@@ -0,0 +1,105 @@
+use std::{cell::UnsafeCell, marker::PhantomData, mem::MaybeUninit, ptr::NonNull};
+
+/// Type-erased pointer into memory. Guaranteed to be correctly aligned, non-null and safe to read for a particular type.
```

Comment:
"Guaranteed" feels too strong to me, since this must be carefully constructed to be safe.

Reply:
The point of these types is that they _are_ supposed to be carefully constructed to uphold these guarantees.